Mathlib/Data/Int/Defs.lean

```diff
@@ -264,6 +264,15 @@ protected theorem le_induction_down {P : ℤ → Prop} {m : ℤ} (h0 : P m)
     fun k hle hi _ ↦ h1 k hle (hi hle)
 #align int.le_induction_down Int.le_induction_down
 
+/-- A strong recursor for `Int` that specifies explicit values for integers below a threshold,
+and is analogous to `Nat.strongRec` for integers on or above the threshold. -/
+@[elab_as_elim] protected def strongRec {P : ℤ → Sort*} (m : ℤ) (lt : ∀ n < m, P n)
+    (ge : ∀ n ≥ m, (∀ k : ℤ, k < n → P k) → P n) (n : ℤ) : P n :=
+  (em <| n < m).by_cases (lt n) fun h ↦ ge n (Int.not_lt.mp h) <| n.inductionOn' m lt
+    (fun _ _ hn l lt1 ↦ (em <| l < m).by_cases (lt l) fun nlt ↦ ge l (Int.not_lt.mp nlt)
+      fun n lt2 ↦ hn n <| Int.lt_of_lt_of_le lt2 <| lt_add_one_iff.mp lt1)
+    fun n _ hn l lt1 ↦ hn l <| Int.lt_trans lt1 n.pred_self_lt
```

Comment:
To me this looks like golfing at its worst, i.e., making sure that if the proof breaks then it is extremely hard to fix :-)
Would it make sense to have a more readable tactic proof? (I'm not sure because this could also be constructing data, and in this case having a term like you do is probably a good idea). An intermediate stance would be to add comments explaining what the different steps are doing.

Reply:
I think term-mode proofs like this is quite hard to break; it only breaks if you remove the lemmas used or change explicitness of arguments, and it's clear how to fix after those changes.

If I had done it in tactic mode I probably would keep it Prop valued rather than Sort*.
 
And I think this falls into the "mathematically trivial" sort of lemma/def that ought to be made as short as possible. Part of the reason this is long is that it's in a basic file that doesn't know Int is a LinearOrder, so some order lemmas with names suitable for dot notation can't be used.